src/linear_algebra/trace.lean

```diff
@@ -96,4 +98,31 @@ if H : ∃ (s : finset M), nonempty (basis s R M) then let ⟨s, ⟨b⟩⟩ := H
 by { simp_rw [trace_eq_matrix_trace R b, linear_map.to_matrix_mul], apply matrix.trace_mul_comm }
 else by rw [trace, dif_neg H, linear_map.zero_apply, linear_map.zero_apply]
 
+/-- The trace of an endomorphism is invariant under conjugation -/
+@[simp]
+theorem trace_conj (g : M →ₗ[R] M) (f : units (M →ₗ[R] M)) :
+  trace R M (↑f * g * ↑f⁻¹) = trace R M g :=
+by { rw trace_mul_comm, simp }
+
+/-- The trace of an endomorphism is invariant under conjugation -/
+@[simp]
+theorem trace_conj' (f g : M →ₗ[R] M) [invertible f] :
+  trace R M (f * g * ⅟ f) = trace R M g :=
+by { rw trace_mul_comm, simp }
+
+end
+
+section
+variables (R : Type u) [field R] {M : Type v} [add_comm_group M] [module R M]
+open_locale classical
+open linear_map
+
+/-- The trace of the identity endomorphism is the dimension of the vector space -/
+@[simp] theorem trace_one : trace R M 1 = finrank R M :=
+if H : ∃ (s : finset M), nonempty (basis s R M) then let ⟨s, ⟨b⟩⟩ := H in
+by { rw [trace_eq_matrix_trace R b, to_matrix_one, finrank_eq_card_finset_basis b], simp }
+else by { rw [trace, dif_neg H, finrank_eq_zero_of_not_exists_basis H], simp }
```

Comment:
If you need classical logic inside a proof, you should use the `classical` tactic rather than `open_locale classical`:
```suggestion

/-- The trace of the identity endomorphism is the dimension of the vector space -/
@[simp] theorem trace_one : trace R M 1 = (finrank R M : R) :=
begin
  classical,
  by_cases H : ∃ (s : finset M), nonempty (basis s R M),
  { obtain ⟨s, ⟨b⟩⟩ := H,
    rw [trace_eq_matrix_trace R b, to_matrix_one, finrank_eq_card_finset_basis b],
    simp, },
  { suffices : (finrank R M : R) = 0, { simp [this, trace, H], },
    simp [finrank_eq_zero_of_not_exists_basis H], },
end
```